src/services/ui/src/api/submissionService.ts

```diff
@@ -0,0 +1,143 @@
+import { API } from "aws-amplify";
+import {OnemacAttachmentSchema, Authority, ReactQueryApiError, Action} from "shared-types";
+import {buildActionUrl, SubmissionServiceEndpoint} from "@/lib";
+import {OneMacUser} from "@/api/useGetUser";
+import {useMutation, UseMutationOptions} from "@tanstack/react-query";
+
+type SubmissionServiceParameters<T> = {
+  data: T,
+  endpoint: SubmissionServiceEndpoint
+  user: OneMacUser | undefined,
+  authority?: Authority
+}
+type SubmissionServiceResponse = {
+  body: {
+    message: string;
+  };
+};
+type PreSignedURL = {
+  url: string;
+  key: string;
+  bucket: string;
+};
+type UploadRecipe = PreSignedURL & {
+  data: File[],
+  title: string
+}
+type AttachmentsSchema = Record<string, File[] | undefined>
+
+/** Pass in an array of UploadRecipes and get a back-end compatible object
+ * to store attachment data */
+const buildAttachmentObject = (
+  recipes: UploadRecipe[]
+): OnemacAttachmentSchema[] => {
+  return recipes.map((r) => r.data.map(f => ({
+    key: r.key,
+    filename: f.name,
+    title: r.title,
+    bucket: r.bucket,
+    uploadDate: Date.now()
+  }) satisfies OnemacAttachmentSchema)).flat();
+};
+
+/** Builds the payload for submission based on which variant a developer has
+ * configured the {@link submit} function with */
+const buildSubmissionPayload = <T extends Record<string, unknown>>(
+  data: T,
+  user: OneMacUser | undefined,
+  endpoint: SubmissionServiceEndpoint,
+  authority?: string,
+  attachments?: UploadRecipe[],
+) => {
+  const userDetails = {
+    submitterEmail: user?.user?.email ?? "N/A",
+    submitterName:
+      `${user?.user?.given_name} ${user?.user?.family_name}` ?? "N/A"
+  };
+  const seaToolFriendlyTimestamp = Math.floor(new Date().getTime() / 1000) * 1000; // Truncating to match seatool
+
+  switch (endpoint) {
+    case "/submit":
+      return {
+        ...data,
+        ...userDetails,
+        proposedEffectiveDate: (data.proposedEffectiveDate as Date).getTime(),
+        attachments: attachments ? buildAttachmentObject(attachments) : null,
+        state: (data.id as string).split("-")[0],
+        authority: authority,
+        origin: "micro",
+      };
+    case buildActionUrl(Action.ISSUE_RAI):
+      return {
+        ...data,
+        ...userDetails,
+        requestedDate: seaToolFriendlyTimestamp,
+        attachments: attachments ? buildAttachmentObject(attachments) : null
+      };
+    case buildActionUrl(Action.RESPOND_TO_RAI):
+      return {
+        ...data,
+        ...userDetails,
+        responseDate: seaToolFriendlyTimestamp,
+        attachments: attachments ? buildAttachmentObject(attachments) : null
+      };
```

Comment:
I like the idea of `switch`-ing on our endpoint value because it can help us identify areas where our payloads are literally the same, but have a renamed attribute here and there. We should strive to keep this switch to as few special cases as possible and group similar payloads. 

For instance, `requestedDate` and `responseDate` don't need to be two fields in two payloads...that's just a timestamp. We can derive whether the timestamp is for a response or request based on the action itself.